src/builder-module.c

```diff
@@ -1276,13 +1299,34 @@ build (GFile          *app_dir,
   g_ptr_array_add (args, g_strdup (argv1));
   while ((arg = va_arg (ap, const gchar *)))
     {
-      if (arg == strv_arg)
+      if (arg == strv_arg || arg == secret_arg)
         {
           argv = va_arg (ap, const gchar **);
           if (argv != NULL)
             {
               for (i = 0; argv[i] != NULL; i++)
-                g_ptr_array_add (args, g_strdup (argv[i]));
+                {
+                  if (arg == secret_arg)
+                    {
+                      const char *cur = strchr (argv[i], '$');
+
+                      if (cur)
+                        {
+                          g_autofree char *secret_key = g_strndup (argv[i], cur - argv[i]);
+                          const char *secret_value = getenv (cur + 1);
```

Comment:
While I think you're definitely on the right path, I'm afraid this still leaks the secret values. That's because a few lines below, these args are passed to `builder_maybe_host_spawnv()`, which calls `flatpak_spawnv()`.

If you look at the implementation of `flatpak_spawnv()` at `src/builder-flatpak-utils.c`, you'll notice that it has the following lines:

```c
  commandline = flatpak_quote_argv ((const char **)argv);
  g_debug ("Running: %s", commandline);
```

This would effectively print the resolved secret arguments :slightly_frowning_face: 

I had a very similar commit locally, and this is exactly where I got stuck at. I think I have an idea on how to continue: make `flatpak_spawnv()` receive an environment (i.e. a list of env vars and values) to set, and never actually resolve the env vars.

Reply:
With maybe too much duplicate code, I make it print a censored version of the command.

Reply:
Note: I'm actually trying something else too.

Reply:
Trying to set them though flatpak env var will leak them:

`FB: Running: flatpak build --die-with-parent --env=FLATPAK_BUILDER_BUILDDIR=/run/build/obs --nofilesystem=host --filesystem=/home/tytan652/Programming/Flatpak/obs-studio/.flatpak-builder/build/obs-3 --bind-mount=/run/build/obs=/home/tytan652/Programming/Flatpak/obs-studio/.flatpak-builder/build/obs-3 --build-dir=/run/build/obs/_flatpak_build --bind-mount=/run/ccache=/home/tytan652/Programming/Flatpak/obs-studio/.flatpak-builder/ccache --env=SOURCE_DATE_EPOCH=1628833344 '--env=CFLAGS=-O2 -g -pipe -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection ' '--env=CXXFLAGS=-O2 -g -pipe -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection ' '--env=LDFLAGS=-L/app/lib -Wl,-z,relro,-z,now -Wl,--as-needed ' --env=CCACHE_DIR=/run/ccache/disabled --env=PATH=/app/bin:/usr/bin --env=LD_LIBRARY_PATH=/app/lib --env=PKG_CONFIG_PATH=/app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig --env=FLATPAK_BUILDER_N_JOBS=4 `**`--env=CEF=/app/cef`**` /home/tytan652/Programming/Flatpak/obs-studio/.flatpak-builder/rofiles/rofiles-Sse630 cmake '-DCMAKE_INSTALL_PREFIX:PATH='\''/app'\''' -G Ninja .. -DCMAKE_BUILD_TYPE=Release -DENABLE_WAYLAND=ON -DBUILD_BROWSER=ON -DUNIX_STRUCTURE=ON -DUSE_XDG=ON -DDISABLE_ALSA=ON -DENABLE_PULSEAUDIO=ON -DWITH_RTMPS=ON -DCEF_ROOT_DIR=$CEF`

It seems that the censor idea, is better.

Reply:
Off-topic, github redacts them if leak in the log.
https://docs.github.com/en/actions/reference/encrypted-secrets#accessing-your-secrets